ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+    /// Panics when the recursion factor is smaller than 2 or masks cannot be set safely (implementation issue).
+    fn compute_proof<C, I>(ctx: &C, uv_tuple_block: I) -> (Self, Self, (Fp61BitPrime, Fp61BitPrime))
+    where
+        C: Context,
+        I: Iterator<Item = UVTupleBlock<Fp61BitPrime>> + Clone,
+    {
+        // set up record counter
+        let mut record_counter = RecordId::from(0);
+
+        // generate first proof from input iterator
+        let (mut uv_values, first_proof_from_left, prover_left_first_proof) = {
+            // precomputation for first proof
+            // first denominator
+            let first_denominator = CanonicalLagrangeDenominator::<
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+            >::new();
+            // first lagrange table
+            let first_lagrange_table = LagrangeTable::<
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+                { LargeProofGenerator::LAGRANGE_LENGTH },
+            >::from(first_denominator);
+
+            LargeProofGenerator::compute_next_proof(
+                ctx,
+                &mut record_counter,
+                &first_lagrange_table,
+                Self::polynomials_from_blocks(uv_tuple_block),
+            )
+        };
+
+        // storage for other proofs
+        let mut prover_left_proofs =
+            Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+        let mut proofs_from_left = Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+
+        // precomputation for other proofs
+        // denominator
+        let denominator = CanonicalLagrangeDenominator::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+        >::new();
+        // lagrange table
+        let lagrange_table = LagrangeTable::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+            { SmallProofGenerator::LAGRANGE_LENGTH },
+        >::from(denominator);
+
+        // recursively generate proofs
+        // until there are less than `RECURSION_FACTOR * (RECURSION_FACTOR-1)` many (u,v) field elements
+        // and then do one more iteration
+        // This ensures that after the last iteration,
+        // there are at most RECURSION_FACTOR-1 many real (u,v) field elements
+        // Therefore at least the last field element is a `F:ZERO` filler that can be replaced
+        // Therefore we have space to include the masks during the final proof
+        //
+        // we remark that uv_value.len() will always be at least 1
+        // further, polynomial uv_values.uv[0].0 and .1 will always have
+        // RECURSION_FACTOR many points
+        // since they are arrays of fixed length, i.e. RECURSION_FACTOR,
+        // (which are filled with `F::ZERO` when there are not enough elements).
+        let r_minus_one = SmallProofGenerator::RECURSION_FACTOR - 1;
```

Comment:
`r` is the name of the challenge point. We are not using `r` to indicate the recursion factor. If you want to do it this way, call it `recursion_factor_minus_one` or `lambda_minus_one`